internal/task/task.go

```diff
@@ -0,0 +1,314 @@
+	// keeping previous time to -N minutes before scheduling interval.
+	task.PreviousRunTime = time.Now().Add(-task.SchedulingInterval)
+	// keeping nextruntime as current time.
+	task.NextRunTime = time.Now()
+	return task
+}
+
+// AddTaskParams methods adds tasks params.
+func (t *Task) AddTaskParams(tp TaskParams) {
+	t.TaskParams = tp
+	t.ID = tp.Config.Name + "$" + tp.Config.RepositoryOwner + "$" + tp.Config.RepositoryName
+}
+
+// AddInterval methods adds tasks scheduling interval.
+func (t *Task) AddInterval(interval time.Duration) {
+	t.SchedulingInterval = interval
+}
+
+// scheduleNextRunOfTask schedules the next run of the task.
+func (task *Task) ScheduleNextRunOfTask() {
+	task.NextRunTime = time.Now().Add(task.SchedulingInterval)
+}
+
+// Start methods starts teh execution of a particular task.
+func (t *Task) Start() error {
+	var ts TaskStats
+	ts, err := getTaskStats(t.ID)
+	// if error reading previous stats , putting default values for task stats map
+	if err != nil {
+		ts.LastIssueTime = t.PreviousRunTime
+		for _, br := range t.Config.Branches {
+			ts.LastCommitTime[br] = t.PreviousRunTime
+		}
+		ts.LastPullRequestNo = 0
+		ts.TaskID = t.ID
+		// saving default task stats for a task.
+		saveTaskStats(t.ID, ts)
+	}
+	gp := gitprovider.NewGitProvider(t.Config.RepositoryHost, t.Config.RepositoryOwner, t.Config.RepositoryName, t.Config.Username, t.DecodeAccessKey)
+
+	// executing for each target in task
+	for _, tar := range t.Targets {
+		// getting new publisher
+		pb, err := publisher.NewPublisher(tar.Type, tar.TargetConfig)
+		if err != nil {
+			log.Errorf("error[%v] in getting publisher for the task with ID %v", err, t.ID)
+			continue
+		}
+		// getting new dataprocessor
+		dp := dataprocessor.NewDataProcessor(t.Config.RepositoryHost, t.Config.RepositoryName, t.Config.RepositoryURL)
+		err = t.collectAndPublishCommits(gp, pb, dp, ts)
+		if err != nil {
+			log.Errorf("error[%v] in collecting commits for task with ID %v", err, t.ID)
+			continue
+		}
+		err = t.collectAndPublishPullRequests(gp, pb, dp, ts)
+		if err != nil {
+			log.Errorf("error[%v] in collecting pull requests for task with ID %v", err, t.ID)
+			continue
+		}
+		err = t.collectAndPublishIssues(gp, pb, dp, ts)
+		if err != nil {
+			log.Errorf("error[%v] in collecting issues for task with ID %v", err, t.ID)
+			continue
+		}
+	}
+	ts, err = getTaskStats(t.ID)
+	if err != nil {
+		log.Errorf("error[%v] in getting task stats for task with ID %v", err, t.ID)
+		return err
+	}
+	ts.LastSuccessFullRunTime = time.Now()
+	// saving task stats.
+	saveTaskStats(t.ID, ts)
+	return nil
+}
+
+//TODO: add stop function in future if required
+func (t *Task) Stop() error {
+	return nil
+}
+
+// collectAndPublishCommits collects commits and publish them to targets.
+func (t *Task) collectAndPublishCommits(gp gitprovider.GitProvider, pb publisher.Publisher, dp dataprocessor.DataProcessor, ts TaskStats) error {
+	var err error
+	for _, br := range t.Config.Branches {
```

Comment:
Instead of running loop, can you run concurrent go routines.

Reply:
sure will add anonymous go func